File: ios/Classes/FlutterMeshNetworkPlugin.swift
import Flutter
import UIKit

/// Main plugin entry point for flutter_mesh_network on iOS.
///
/// Registers two MethodChannels and delegates calls to dedicated handlers:
/// - `flutter_mesh_network/nearby` -> MeshConnectivityHandler (MultipeerConnectivity)
/// - `flutter_mesh_network/ble`    -> BlePeripheralHandler   (CoreBluetooth)
public class FlutterMeshNetworkPlugin: NSObject, FlutterPlugin {

    private let meshHandler = MeshConnectivityHandler()
    private let bleHandler = BlePeripheralHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterMeshNetworkPlugin()

        // --- Nearby (MultipeerConnectivity) channel ---
        let nearbyChannel = FlutterMethodChannel(
            name: "flutter_mesh_network/nearby",
            binaryMessenger: registrar.messenger()
        )
        instance.meshHandler.setMethodChannel(nearbyChannel)
        nearbyChannel.setMethodCallHandler(instance.handleNearbyCall)

        // --- BLE Peripheral channel ---
        let bleChannel = FlutterMethodChannel(
            name: "flutter_mesh_network/ble",
            binaryMessenger: registrar.messenger()
        )
        instance.bleHandler.setMethodChannel(bleChannel)
        bleChannel.setMethodCallHandler(instance.handleBleCall)

        // Keep the instance alive for the duration of the registrar.
        registrar.addMethodCallDelegate(instance, channel: nearbyChannel)
    }

    // MARK: - FlutterPlugin (required but unused — routing happens via per-channel handlers)

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // This is only called for the nearbyChannel because of addMethodCallDelegate.
        // We route through handleNearbyCall instead, so this is a no-op fallback.
        result(FlutterMethodNotImplemented)
    }

    // MARK: - Nearby Channel Handler

    private func handleNearbyCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "isAvailable":
            result(meshHandler.isAvailable())

        case "publish":
            let serviceName = args["serviceName"] as? String ?? "mesh-net"
            let userId      = args["userId"] as? String ?? ""
            let userName    = args["userName"] as? String ?? ""
            let latitude    = args["latitude"] as? Double
            let longitude   = args["longitude"] as? Double
            result(meshHandler.startPublishing(
                serviceName: serviceName,
                userId: userId,
                userName: userName,
                latitude: latitude,
                longitude: longitude
            ))

        case "subscribe":
            let serviceName = args["serviceName"] as? String ?? "mesh-net"
            result(meshHandler.startSubscribing(serviceName: serviceName))

        case "sendMessage":
            let peerId = args["peerId"] as? String ?? ""
            let data   = args["data"] as? String ?? ""
            result(meshHandler.sendMessage(peerId: peerId, data: data))

        case "broadcastMessage":
            let data = args["data"] as? String ?? ""
            result(meshHandler.broadcastMessage(data: data))

        case "measureDistance":
            let peerId = args["peerId"] as? String ?? ""
            if let distance = meshHandler.measureDistance(peerId: peerId) {
                result(distance)
            } else {
                result(nil)
            }

        case "stopPublish":
            meshHandler.stopPublish()
            result(true)

        case "stopSubscribe":
            meshHandler.stopSubscribe()
            result(true)

        case "updateLocation":
            let latitude  = args["latitude"] as? Double ?? 0
            let longitude = args["longitude"] as? Double ?? 0
            meshHandler.updateLocation(latitude: latitude, longitude: longitude)
            result(true)

        case "getPeers":
            result(meshHandler.getPeers())

        case "stop":
            meshHandler.stop()
            result(true)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - BLE Channel Handler

    private func handleBleCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "initialize":
            result(bleHandler.initialize())

        case "startGattServer":
            result(bleHandler.startGattServer())

        case "startAdvertising":
            let name = args["userName"] as? String ?? ""
            let lat  = args["latitude"] as? Double
            let lng  = args["longitude"] as? Double
            result(bleHandler.startAdvertising(name: name, lat: lat, lng: lng))

        case "stopAdvertising":
            bleHandler.stopAdvertising()
            result(true)

        case "updateLocation":
            let lat = args["latitude"] as? Double
            let lng = args["longitude"] as? Double
            bleHandler.updateAdvertising(lat: lat, lng: lng)
            result(true)

        case "notifyAll":
            let data = args["data"] as? String ?? ""
            let payload = data.data(using: .utf8) ?? Data()
            result(bleHandler.notifyAllSubscribers(data: payload))

        case "getState":
            result(bleHandler.getState())

        case "stop":
            bleHandler.stop()
            result(true)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/fluttermeshnetwork/flutter_mesh_network/FlutterMeshNetworkPlugin.kt
package com.fluttermeshnetwork.flutter_mesh_network

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodChannel

/**
 * FlutterMeshNetworkPlugin — main entry point for the flutter_mesh_network plugin.
 *
 * Registers two MethodChannels:
 *   - flutter_mesh_network/nearby  → Wi-Fi Aware (NAN)
 *   - flutter_mesh_network/ble     → BLE Peripheral (GATT server + advertising)
 *
 * Delegates all calls to [WifiAwareHandler] and [BlePeripheralHandler] respectively.
 */
class FlutterMeshNetworkPlugin : FlutterPlugin, ActivityAware {

    private var nearbyChannel: MethodChannel? = null
    private var bleChannel: MethodChannel? = null

    private var wifiAwareHandler: WifiAwareHandler? = null
    private var blePeripheralHandler: BlePeripheralHandler? = null

    private var activity: Activity? = null

    // =========================================================================
    // FlutterPlugin
    // =========================================================================

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val messenger = binding.binaryMessenger
        val context = binding.applicationContext

        // Wi-Fi Aware channel
        nearbyChannel = MethodChannel(messenger, "flutter_mesh_network/nearby")
        wifiAwareHandler = WifiAwareHandler(context)
        wifiAwareHandler!!.setMethodChannel(nearbyChannel!!)
        nearbyChannel!!.setMethodCallHandler { call, result ->
            wifiAwareHandler!!.onMethodCall(call, result)
        }

        // BLE Peripheral channel
        bleChannel = MethodChannel(messenger, "flutter_mesh_network/ble")
        blePeripheralHandler = BlePeripheralHandler(context)
        blePeripheralHandler!!.setMethodChannel(bleChannel!!)
        bleChannel!!.setMethodCallHandler { call, result ->
            blePeripheralHandler!!.onMethodCall(call, result)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        nearbyChannel?.setMethodCallHandler(null)
        nearbyChannel = null

        bleChannel?.setMethodCallHandler(null)
        bleChannel = null

        wifiAwareHandler?.dispose()
        wifiAwareHandler = null

        blePeripheralHandler?.stop()
        blePeripheralHandler = null
    }

    // =========================================================================
    // ActivityAware
    // =========================================================================

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        wifiAwareHandler?.setActivity(activity)
        blePeripheralHandler?.setActivity(activity)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        wifiAwareHandler?.setActivity(activity)
        blePeripheralHandler?.setActivity(activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        wifiAwareHandler?.setActivity(null)
        blePeripheralHandler?.setActivity(null)
    }

    override fun onDetachedFromActivity() {
        activity = null
        wifiAwareHandler?.setActivity(null)
        blePeripheralHandler?.setActivity(null)
    }
}
